core/src/unprocessed_packet_batches.rs

```diff
@@ -0,0 +1,1205 @@
+                    }
+                });
+        },
+    );
+
+    (stakes, locators)
+}
+
+fn weighted_shuffle(
+    stakes: &[u64],
+    locators: &[PacketLocator],
+    packet_sender_info: &mut Option<PacketSenderInfo>,
+) -> Vec<PacketLocator> {
+    let need_to_shuffle_sender_ips = packet_sender_info.is_some();
+    let mut shuffled_packet_senders_ip = Vec::<IpAddr>::new();
+
+    let mut rng = rand::thread_rng();
+    let shuffled_locators: Vec<PacketLocator> = WeightedShuffle::new(stakes)
+        .unwrap()
+        .shuffle(&mut rng)
+        .map(|i| {
+            if need_to_shuffle_sender_ips {
+                let packet_sender_info = packet_sender_info.as_ref().unwrap();
+                shuffled_packet_senders_ip.push(packet_sender_info.packet_senders_ip[i]);
+            }
+            locators[i].clone()
+        })
+        .collect();
+    if let Some(packet_sender_info) = packet_sender_info {
+        packet_sender_info.packet_senders_ip = shuffled_packet_senders_ip;
+    }
+    shuffled_locators
+}
+
+pub fn prioritize_by_fee(
+    unprocessed_packet_batches: &UnprocessedPacketBatches,
+    locators: &Vec<PacketLocator>,
+    bank: Option<Arc<Bank>>,
+) -> Vec<PacketLocator> {
+    let mut fee_buckets = BTreeMap::<u64, Vec<PacketLocator>>::new();
+    for locator in locators {
+        let fee_per_cu = if let Some(fee_per_cu) =
+            compute_fee_per_cu(unprocessed_packet_batches, locator, &bank)
+        {
+            fee_per_cu
+        } else {
+            // if unable to compute fee-per-cu for the packet, put it to the `0` bucket
+            0u64
+        };
+
+        let bucket = fee_buckets
+            .entry(fee_per_cu)
+            .or_insert(Vec::<PacketLocator>::new());
+        bucket.push(locator.clone());
+    }
+    fee_buckets
+        .iter()
+        .rev()
+        .flat_map(|(_key, bucket)| bucket.iter().cloned())
+        .collect()
+}
+
+// to comute (addition_fee + base_fee / requested_cu) for packet identified by `locator`
+fn compute_fee_per_cu(
+    unprocessed_packet_batches: &UnprocessedPacketBatches,
+    locator: &PacketLocator,
+    bank: &Option<Arc<Bank>>,
+) -> Option<u64> {
+    if let Some(bank) = bank {
+        let deserialized_packet_batch = unprocessed_packet_batches.get(locator.batch_index)?;
+        let deserialized_packet = deserialized_packet_batch
+            .unprocessed_packets
+            .get(&locator.packet_index)?;
+        let sanitized_message = sanitize_message(
+            &deserialized_packet.versioned_transaction.message,
+            bank.as_ref(),
+        )?;
+        let total_fee = bank.get_fee_for_message(&sanitized_message)?;
+
+        // TODO update bank to get_fee_and_cu_for_message() to avoid calling compute_budget again
+        let mut compute_budget = ComputeBudget::new(false);
+        let _ = compute_budget
+            .process_message(&sanitized_message, false)
+            .ok()?;
```

Comment:
What happens if `process_message()` errors here? Then `compute_budget.max_units` will be set to the default which may change under us. We should probably explicitly handle that case here